libass/ass_render.c

```diff
@@ -1185,93 +1347,56 @@ transform_3d(ASS_Vector shift, ASS_Outline *outline, int n_outlines,
  * Tries to get glyph bitmaps from bitmap cache.
  * If they can't be found, they are generated by rotating and rendering the glyph.
  * After that, bitmaps are added to the cache.
- * They are returned in info->bm (glyph), info->bm_o (outline) and info->bm_s (shadow).
+ * They are returned in info->image (glyph), info->image_o (outline).
  */
 static void
 get_bitmap_glyph(ASS_Renderer *render_priv, GlyphInfo *info)
 {
-    if (!info->outline || info->symbol == '\n' || info->symbol == 0 || info->skip)
+    if (!info->outline || info->symbol == '\n' || info->symbol == 0 || info->skip) {
+        ass_cache_dec_ref(info->outline);
+        ass_cache_dec_ref(info->border);
```

Comment:
I haven’t managed to figure out why these calls to `ass_cache_dec_ref` were added. Did we have a leak before, or is this caused by some (which?) of the changes in this commit?

Reply:
I've added that just to be safe. `get_bitmap_glyph()` should always consume outlines if present. I'm not sure if any of the conditions after `!info->outline` can be triggered at all and not handled elsewhere. At least I've seen handling of skip glyphs at `render_and_combine_glyphs()` bypassing call to `get_bitmap_glyph()`.

Reply:
`get_bitmap_glyph()` should either free outline or transfer it's ownership to corresponding bitmap. So it's definitely not an error to free outlines here. But maybe the whole if can be replaced with `if (!info->outline) return;`

Reply:
I've separated that in different commit.